nntrainer/include/network_graph.h

```diff
@@ -0,0 +1,81 @@
+// SPDX-License-Identifier: Apache-2.0
+/**
+ * Copyright (C) 2020 Jijoong Moon <jijoong.moon@samsung.com>
+ *
+ * @file    network_graph.h
+ * @date    19 Oct 2020
+ * @see     https://github.com/nnstreamer/nntrainer
+ * @author  Jijoong Moon <jijoong.moon@samsung.com>
+ * @bug     No known bugs except for NYI items
+ * @brief   This is Network Graph Class for Neural Network
+ *
+ */
+
+#ifndef __NETWORK_GRAPH_H_
+#define __NETWORK_GRAPH_H_
+#ifdef __cplusplus
+
+#include <iostream>
+#include <layer_internal.h>
+#include <list>
+#include <loss_layer.h>
+#include <memory>
+#include <stack>
+#include <vector>
+
+namespace nntrainer {
+
+struct LayerNode {
+  std::shared_ptr<Layer> layer;
+  unsigned int index;
+};
+
+class NetworkGraph {
+public:
+  NetworkGraph() : num_node(0), def_name_count(0){};
+
+  void addEdge(unsigned int ith, LayerNode node);
+
+  void topologicalSort();
+
+  void addLayerNode(std::shared_ptr<Layer>);
+
+  unsigned int getNumNode() { return num_node; }
+
+  LayerNode &getLayerNode(unsigned int ith);
+
+  LayerNode &getLayerNode(const std::string layer_name);
+
+  void ensureName(std::shared_ptr<Layer> layer, const std::string &prefix,
+                  bool force_rename);
+
+  int setGraphNode(std::vector<std::shared_ptr<Layer>> layers,
+                   const LossType loss_type);
+
+  int realizeMultiInputType(Layer &current);
+
+  int realizeActivationType(Layer &current,
+                            std::vector<std::shared_ptr<Layer>> layers);
+
+  int realizeFlattenType(Layer &current,
+                         std::vector<std::shared_ptr<Layer>> layers);
+
+  int addLossLayer(const LossType loss_type);
+
+  int setEdge();
+
+private:
+  void topologicalSortUtil(unsigned int ith, bool visited[],
+                           std::stack<LayerNode> &Stack);
+
+  unsigned int num_node;
+  std::vector<std::list<LayerNode>> adj;
```

Comment:
@jijoongmoon I think this is important to consider at start as this would require big changes, but how about using `std::unordered_map<LayerNode, std::list<LayerNode>>` for adj? This makes searching and traversing the graph easier. You can also consider `std::unordered_map<std::string, std::list<LayerNode>>` where string in the layer name.

Reply:
Yes. that's good idea. but I think we can use it after we complete the whole processes.

Reply:
Changing from vector to unordered_map would require big changes, so I think its much easier to switch as early as possible.